### ISSUE TITLE
Windows Update Rework: Todo list consolidating tickets

### ISSUE THREAD

**Issue by silversword411** [amidaware/tacticalrmm#1188]
Have a TacticalRMM Global Option like: [] Have Tactical RMM manage all Windows Update functions
Block specific patches agent/site/client
Agent/Site/Client Button: "Approve patches based on policy now"
Agent/Site/Client Button: "Approve and install Now"
Run Updates on offline agents if missed
Have Update install Window (time window, with working hours like windows has)
different installation and reboot times
only reboot when accepted by user
postpone updates for x days so it can be tested first
Summary screen on patch status for machines
Manual Mass approve Updates (add automation policy selection to Bulk Patch Management dialog)
Add more time options to scheduled patching e.g. "first, second, Last <weekday> of Month
Run script before or after patching

Replication of all the features in PSWindowsUpdate will probably make everyone happy: <https://adamtheautomator.com/pswindowsupdate/>

**Comment by emtek-at** [comment]
Please also add an ability to schedule patch installation/approval based on severity. So critical patches could be installed on a daily base and e.g. other only once a month.

**Comment by yourITMatters** [comment]
It would be nice if TRMM would run a patch scan at next check in after the patch window and patch / install approved patches based on the the configured automation policy.  that way if a workstation is off during the patching window it will pick up approved patches the next time it is online.  (posting from discord)

**Comment by Answerphones** [comment]
Ability to start the pending updates as soon as the endpoint workstation returns online.
Ability to have all the available patches in a global patch list and approve globally based on patch
Re-work the "other"category by having TRMM split off the security updates (daily) to auto approve and hold for admin anything else.

**Comment by silversword411** [comment]
Allow Optional quality updates

To include "Optional Quality Updates" in your search results with the Windows Update Agent (WUA) API, you need to use the IUpdateSearcher interface to construct a search query that specifically looks for these types of updates. Optional quality updates typically include updates that are not critical for the security of the system but may provide enhancements or non-critical fixes.

Here's a basic example of how to structure your code to search for optional updates using the WUA API in a COM-aware programming environment, such as C# with appropriate COM interop, or VBScript. The example focuses on how to set up the search criteria to include optional updates:

Example in C# (Assuming COM Interop setup)
```
csharp
Copy code
// First, create an instance of the UpdateSession COM object
UpdateSession updateSession = new UpdateSession();
IUpdateSearcher updateSearcher = updateSession.CreateUpdateSearcher();

// Set the search criteria to include optional updates
// This might include updates not automatically downloaded and installed
string searchCriteria = "IsInstalled=0 AND Type='Software' AND IsHidden=0";

// Perform the search
ISearchResult searchResult = updateSearcher.Search(searchCriteria);

// Iterate through the results and process them as needed
foreach (IUpdate update in searchResult.Updates)
{
    if (update.IsOptional)
    {
        Console.WriteLine($"Optional update found: {update.Title}");
        // Further processing, such as displaying to the user for manual selection
    }
}
```
Example in VBScript
```
vbscript
Copy code
' Create the UpdateSession Object
Set updateSession = CreateObject("Microsoft.Update.Session")
Set updateSearcher = updateSession.CreateUpdateSearcher()

' Set the search criteria to include optional updates
searchCriteria = "IsInstalled=0 AND Type='Software' AND IsHidden=0"

' Perform the search
Set searchResult = updateSearcher.Search(searchCriteria)

' Iterate through the search results
For Each update In searchResult.Updates
    If update.IsOptional Then
        WScript.Echo "Optional update found: " & update.Title
        ' Further processing here
    End If
Next
```

These examples use a basic search criteria string that looks for software updates that are not already installed and are not hidden. The IsOptional property check in the loop (applicable in environments like C# with proper property exposure) is a simplified way of identifying updates that might be considered optional. However, it's important to note that the exact way to identify optional quality updates can depend on the specific properties and categories used by the updates at the time of your query.

**Comment by technoicon** [comment]
> Ability to start the pending updates as soon as the endpoint workstation returns online. Ability to have all the available patches in a global patch list and approve globally based on patch Re-work the "other"category by having TRMM split off the security updates (daily) to auto approve and hold for admin anything else.

With this, while I agree it should update automatically when a pc comes online, a delay might be best.  Running windows updates when a computer first starts is not ideal.  I've seen it grind PC's to a holt as windows update service is a cpu and memory hog, if this software is doing it, expect it(tacticalrmm) to be removed and problem "seemingly" solved.

This can give a very bad user experience, especially when they are logging on from a long time away.  This would give unnecessary calls to IT helpdesk. (back from holidays for example)

**Comment by styx-tdo** [comment]
* create a TEST group (by setting an attribute like "canary" or "test group"; boxes in there get updates as soon as they are released and forced. If these exist in a site/customer, the other machines only should get the updates a week/a few days after.

**Comment by joeldeteves** [comment]
@styx-tdo that is technically something that is better handled by InTune / GPO + Windows Update Rings imo unless you don't want to push out security patches during regular patch cycling for some reason?

**Comment by silversword411** [comment]
Personally managing my windows should not require feudal tithes to be paid to manage my system.

Replicate every feature of Intune, and not have to pay a monthly pound of flesh in monthly dues.

**Comment by joeldeteves** [comment]
I'm not so sure Windows would obey your request unless the registry is modified directly. Take for example feature updates - we usually do defer those via InTune. I don't know how you would do that via TRMM, but I'm sure a reg key exists.

I also don't know how you would prevent some machines from installing patches unless you straight up disable automatic updates. I suppose a lot of organizations still do that, but I'm not sure it's a good idea to have the agent itself should be making these types of modifications.

We use InTune and TRMM together to come up with a patch strategy that works for us.

I'm sure everyone has different use cases.

**Comment by technoicon** [comment]
I'm finding Intune a little difficult to do reporting on patch status.
Also deploying patches and the like.
I'm sure I'll work it out but I just wish it was simpler.

Then we have smaller clients 1-2 endpoints that won't pay for it.

Need a patching solution that just keeps things up to date.
Can exclude things like drivers.
And exclude a certain patch across all machines (for those patches that
break stuff).

**Comment by weilix65** [comment]
Please accelerate this update for TRMM because it's really important.
Thank you :-)
BTW:  TRMM is a great help for me.
